pkg/auth/providers/keycloakoidc/keycloak_client.go

```diff
@@ -0,0 +1,179 @@
+			return nil, nil
+		}
+		if err := json.Unmarshal(b, &groups); err != nil {
+			logrus.Errorf("[keycloak oidc]: received error unmarshalling search results, err: %v", err)
+			return accounts, err
+		}
+		for _, g := range groups {
+			accounts = append(accounts, Account{ID: g.ID, Name: g.Name, Type: GroupType})
+			subGroups := getSubGroups(g)
+			for _, sg := range subGroups {
+				accounts = append(accounts, Account{ID: sg.ID, Name: sg.Name, Type: GroupType})
+			}
+		}
+	}
+	return accounts, nil
+}
+
+func getSubGroups(group Group) []Group {
+	var groups []Group
+	if len(group.Subgroups) > 0 {
+		for i, sub := range group.Subgroups {
+			// setting an upper limit for how many subgroups we will loop through
+			// this value was chosen at random so can be changed if needed
+			if i < 100 {
+				groups = append(groups, sub)
+				groups = append(groups, getSubGroups(sub)...)
+			}
+		}
+	}
+	return groups
+}
+
+func (k *KClient) getFromKeyCloakByID(principalID, accessToken, searchType string, config *v32.OIDCConfig) (Account, error) {
+	sURL, err := k.getSearchURL(config.Issuer)
+	if err != nil {
+		return Account{}, nil
+	}
+	searchURL := fmt.Sprintf("%s/%s/%s", sURL, searchType, principalID)
+	search := URLEncoded(searchURL)
+	b, err := k.getFromKeyCloak(accessToken, search, config.Certificate, config.PrivateKey)
+	if err != nil {
+		return Account{}, nil
+	}
+	var searchResult Account
+	if err := json.Unmarshal(b, &searchResult); err != nil {
+		logrus.Errorf("[keycloak oidc]: received error unmarshalling search results, err: %v", err)
+		return searchResult, err
+	}
+	return searchResult, nil
+
+}
+func (k *KClient) getSearchURL(issuer string) (string, error) {
+	splitIssuer := strings.SplitAfter(issuer, "/auth/")
+	return fmt.Sprintf(
+		"%sadmin/%s",
+		splitIssuer[0],
+		splitIssuer[1],
+	), nil
+}
+
+//URLEncoded encodes the string
+func URLEncoded(str string) string {
+	u, err := url.Parse(str)
+	if err != nil {
+		logrus.Errorf("Error encoding the url: %s, error: %v", str, err)
+		return str
+	}
+	return u.String()
+}
+
+func (k *KClient) getFromKeyCloak(accessToken, url, certificate, key string) ([]byte, error) {
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return nil, err
+	}
+	err = common.GetClientWithCertKey(k.httpClient, certificate, key)
+	if err != nil {
+		return nil, err
+	}
+	req.Header.Add("Authorization", "token "+accessToken)
+	req.Header.Add("Accept", "application/json")
+	resp, err := k.httpClient.Do(req)
+	if err != nil {
+		logrus.Errorf("[keycloak oidc] received error from keycloak: %v", err)
+		return nil, err
+	}
+	defer resp.Body.Close()
+	// Check the status code
+	switch resp.StatusCode {
+	case 200:
+	case 201:
+	default:
+		var body bytes.Buffer
+		io.Copy(&body, resp.Body)
+		return nil, fmt.Errorf("[keycloak oidc] request failed, got status code: %d. Response: %s",
+			resp.StatusCode, body.Bytes())
+	}
+	b, err := ioutil.ReadAll(resp.Body)
+	return b, err
```

Comment:
Just return `b, err := ioutil.ReadAll(resp.Body)` and let the called handler the data and the error.

Reply:
I have changed the code to just return the body, error code and error now and handle the additional checks where I call this method